Android Notifications: When, ShowWhen, Color

dahlbyk opened the pull request:
### Description of Change ###

Add `AndroidOptions` properties for:

- [`when`](https://developer.android.com/reference/android/app/Notification#when) ([`setWhen`](https://developer.android.com/reference/android/app/Notification.Builder.html#setWhen(long))/[`setShowWhen`](https://developer.android.com/reference/android/app/Notification.Builder.html#setShowWhen(boolean)))
- [`color`](https://developer.android.com/reference/android/app/Notification.html#color) ([`setColor`](https://developer.android.com/reference/android/app/Notification.Builder.html#setColor(int)))

### API Changes ###
<!-- List all API changes here (or just put None) -->
 
- `Notifications.AndroidOptions` property changes:
  - Add `bool? ShowWhen`, `DateTime? When`, and `static bool? DefaultShowWhen`
  - Add `string? ColorResourceName` and `static string? DefaultColorResourceName`
  - Mark `ChannelDescription` and `DefaultChannelDescription` as nullable

### Platforms Affected ### 

- Android

### Behavioral Changes ###

Notifications should be created with the given properties.

~~**However:** The color `Resources` lookup doesn't seem to work. Not sure what I'm doing wrong? Or is using a resource the wrong approach?~~ No idea what I was seeing last night, but it's working now. 🤷‍♂ 

### Testing Procedure ###

Verified in my app.

### PR Checklist ###

- [x] Rebased on top of the target branch at time of PR
- [ ] Changes adhere to coding standard

dahlbyk commented:
> Should missing Notification resource be non-fatal?

Added this commit to avoid a runtime error for something with reasonable default behavior; glad to revert if you prefer.

aritchie commented on src/Shiny.Notifications/Platforms/Android/NotificationManager.cs:
```diff
@@ -217,14 +228,32 @@ protected virtual PendingIntent GetLaunchPendingIntent(Notification notification
+            var colorResourceId = this.context.GetColorByName(colorResourceName);
+            if (colorResourceId <= 0)
+            {
+                Log.Write(NotificationLogCategory.Notifications, $"Color ResourceId for {colorResourceName} not found");
```
Always throw on bad args from the user - otherwise they tend to file bugs saying it isn't working, where in actual fact, it isn't finding the resource because they haven't set it up

aritchie commented on src/Shiny.Notifications/Platforms/Android/NotificationManager.cs:
```diff
@@ -217,14 +228,32 @@ protected virtual PendingIntent GetLaunchPendingIntent(Notification notification
 
             var smallIconResourceId = this.context.GetResourceIdByName(notification.Android.SmallIconResourceName);
             if (smallIconResourceId <= 0)
-                throw new ArgumentException($"Icon ResourceId for {notification.Android.SmallIconResourceName} not found");
```
Same issue as color resourceId

aritchie commented on src/Shiny.Notifications/Platforms/Android/NotificationManager.cs:
```diff
@@ -253,11 +282,11 @@ protected virtual void AddCategory(NotificationCompat.Builder builder, Notificat
             if (category == null)
             {
-                Log.Write("Notifications", "No notification category found for " + notification.Category);
+                Log.Write(NotificationLogCategory.Notifications, "No notification category found for " + notification.Category);
```
To understand why this is different than the resourceId exceptions above, it is because the user can change there categories from one day to the next and all scheduled notifications would start to fail.  Categories are not as "static" as resources and general purpose args

dahlbyk commented:
> > Should missing Notification resource be non-fatal?
> 
> Added this commit to avoid a runtime error for something with reasonable default behavior; glad to revert if you prefer.

Reverted.

Also refactored the API level check to align with the pattern you used in 550d8c34b24380ca8394e85f5dec03ae4e4d7068.

aritchie commented:
Thanks for the PR
